app/src/main/java/de/westnordost/streetcomplete/quests/housenumber/AddHousenumber.kt

```diff
@@ -38,111 +31,90 @@ class AddHousenumber(private val overpass: OverpassMapDataAndGeometryApi) : OsmE
+        val addressNodeIds = addressNodesById.keys
 
-        val buildings = downloadBuildingsWithoutAddresses(bbox) ?: return false
-        // empty result: We are done
-        if (buildings.isEmpty()) return true
+        /** filter: only buildings with no address that usually should have an address
+         *          ...that do not have an address node on their outline */
 
-        val addrAreas = downloadAreasWithAddresses(bbox) ?: return false
+        val buildings = mapData.filter {
+            buildingsWithMissingAddressFilter.matches(it)
+            && !it.containsAnyNode(addressNodeIds, mapData)
+        }.toMutableList()
 
-        val extendedBbox = getBoundingBoxThatIncludes(buildings.values)
+        if (buildings.isEmpty()) return listOf()
 
-        val addrPositions = downloadFreeFloatingPositionsWithAddresses(extendedBbox) ?: return false
+        /** exclude buildings which are included in relations that have an address */
 
-        Log.d("AddHousenumber", "Downloaded ${buildings.size} buildings with no address, " +
-            "${addrAreas.size} areas with address and ${addrPositions.size} address nodes " +
-            "in ${System.currentTimeMillis() - ms}ms"
-        )
-        ms = System.currentTimeMillis()
+        val relationsWithAddress = mapData.relations.filter { nonMultipolygonRelationsWithAddressFilter.matches(it) }
 
-        val buildingPositions = LatLonRaster(extendedBbox, 0.0005)
-        for (buildingCenter in buildings.keys) {
-            buildingPositions.insert(buildingCenter)
+        buildings.removeAll { building ->
+            relationsWithAddress.any { it.containsWay(building.id) }
         }
 
-        // exclude buildings that are contained in an area with a housenumber
-        for (addrArea in addrAreas) {
-            for (buildingPos in buildingPositions.getAll(addrArea.getBounds())) {
-                if (buildingPos.isInMultipolygon(addrArea.polygons)) {
-                    buildings.remove(buildingPos)
-                }
-            }
-        }
+        if (buildings.isEmpty()) return listOf()
 
-        var createdQuests = 0
-        // only buildings with no housenumber-nodes inside them
-        for (building in buildings.values) {
-            // even though we could continue here, limit the max amount of quests created to the
-            // default maximum to avoid performance problems
-            if (createdQuests++ >= DEFAULT_MAX_QUESTS) break
-
-            val addrContainedInBuilding = getPositionContainedInBuilding(building.geometry, addrPositions)
-            if (addrContainedInBuilding != null) {
-                addrPositions.remove(addrContainedInBuilding)
-                continue
-            }
+        /** exclude buildings that intersect with the bounding box because it is not possible to
+           ascertain for these if there is an address node within the building - it could be outside
+           the bounding box */
```

Comment:
Unfortunately, the problem here is that in small bounding boxes, quite a few buildings will be excluded because of this. If a housenumber quest was downloaded before as part of a big bounding box and then another small download is being executed, housenumber quests on buildings on the edge will vanish.

The prior solution was to expand the bounding box to include all the buildings and then download the housenumber nodes based on that enlarged bounding box. With the solution now, this is not possible (or: not expedient) anymore - we have that one bounding box with the data and we should create all the quests based on the downloaded data.